File: experiments/bakeoff/benchmarks/python-inventory/ground_truth/reference/shop/services/inventory_service.py
"""services/inventory_service.py — Product catalog + stock + reservations.

Reservation lifecycle (one token per (order, sku) line):
  reserve()  -> opens token, increments StockLevel.reserved
  commit()   -> closes token, decrements on_hand AND reserved
  release()  -> closes token, decrements reserved (returns to available)
A token can be committed XOR released, never both, and never twice.
"""

from ..errors import (
    ConflictError,
    InsufficientStockError,
    NotFoundError,
    ReservationError,
    ValidationError,
)
from ..persistence import Store
from ..types.inventory import ReservationToken, StockLevel
from ..types.products import Product


class InventoryService:
    def __init__(self, store: Store) -> None:
        self.store = store
        self._token_seq = 0

    def register_product(self, *, sku: str, name: str, price: float) -> Product:
        if sku in self.store.products:
            raise ConflictError(f"product with sku {sku} already exists")
        product = Product(sku=sku, name=name, price=price)
        self.store.products[sku] = product
        self.store.stock[sku] = StockLevel(sku=sku)
        return product

    def get_product(self, sku: str) -> Product:
        product = self.store.products.get(sku)
        if product is None:
            raise NotFoundError(f"product not found: {sku}")
        return product

    def stock_of(self, sku: str) -> StockLevel:
        stock = self.store.stock.get(sku)
        if stock is None:
            raise NotFoundError(f"no stock record for sku {sku}")
        return stock

    def add_stock(self, sku: str, qty: int) -> None:
        if qty <= 0:
            raise ValidationError("add_stock qty must be > 0")
        stock = self.stock_of(sku)
        stock.on_hand += qty

    def reserve(self, *, order_id: str, sku: str, quantity: int) -> ReservationToken:
        if quantity <= 0:
            raise ValidationError("reserve quantity must be > 0")
        stock = self.stock_of(sku)
        if stock.available < quantity:
            raise InsufficientStockError(
                f"insufficient stock for {sku}: have {stock.available}, need {quantity}"
            )
        stock.reserved += quantity
        self._token_seq += 1
        token_id = f"rsv-{self._token_seq:06d}"
        token = ReservationToken(
            token_id=token_id,
            order_id=order_id,
            sku=sku,
            quantity=quantity,
        )
        self.store.reservations[token_id] = token
        return token

    def commit(self, token_id: str) -> None:
        token = self.store.reservations.get(token_id)
        if token is None:
            raise NotFoundError(f"reservation not found: {token_id}")
        if not token.is_open:
            raise ReservationError(f"reservation {token_id} already closed")
        stock = self.stock_of(token.sku)
        stock.on_hand -= token.quantity
        stock.reserved -= token.quantity
        token.committed = True

    def release(self, token_id: str) -> None:
        token = self.store.reservations.get(token_id)
        if token is None:
            raise NotFoundError(f"reservation not found: {token_id}")
        if not token.is_open:
            raise ReservationError(f"reservation {token_id} already closed")
        stock = self.stock_of(token.sku)
        stock.reserved -= token.quantity
        token.released = True

File: experiments/bakeoff/benchmarks/dart-inventory/ground_truth/reference/lib/services/inventory_service.dart
/// services/inventory_service.dart — Product catalog + stock + reservations.
///
/// Reservation lifecycle (one token per (order, sku) line):
///   reserve()  -> opens token, increments StockLevel.reserved
///   commit()   -> closes token, decrements onHand AND reserved
///   release()  -> closes token, decrements reserved (returns to available)
/// A token can be committed XOR released, never both, and never twice.

import '../errors.dart';
import '../persistence.dart';
import '../types/products.dart';
import '../types/inventory.dart';

class InventoryService {
  final Store store;
  int _tokenSeq = 0;

  InventoryService(this.store);

  Product registerProduct({
    required String sku,
    required String name,
    required double price,
  }) {
    if (store.products.containsKey(sku)) {
      throw ConflictError('product with sku $sku already exists');
    }
    final p = Product(sku: sku, name: name, price: price);
    store.products[sku] = p;
    store.stock[sku] = StockLevel(sku: sku);
    return p;
  }

  Product getProduct(String sku) {
    final p = store.products[sku];
    if (p == null) {
      throw NotFoundError('product not found: $sku');
    }
    return p;
  }

  StockLevel stockOf(String sku) {
    final s = store.stock[sku];
    if (s == null) {
      throw NotFoundError('no stock record for sku $sku');
    }
    return s;
  }

  void addStock(String sku, int qty) {
    if (qty <= 0) {
      throw ValidationError('addStock qty must be > 0');
    }
    final s = stockOf(sku);
    s.onHand += qty;
  }

  ReservationToken reserve({
    required String orderId,
    required String sku,
    required int quantity,
  }) {
    if (quantity <= 0) {
      throw ValidationError('reserve quantity must be > 0');
    }
    final s = stockOf(sku);
    if (s.available < quantity) {
      throw InsufficientStockError(
        'insufficient stock for $sku: have ${s.available}, need $quantity',
      );
    }
    s.reserved += quantity;
    final tokenId = 'rsv-${(++_tokenSeq).toString().padLeft(6, '0')}';
    final token = ReservationToken(
      tokenId: tokenId,
      orderId: orderId,
      sku: sku,
      quantity: quantity,
    );
    store.reservations[tokenId] = token;
    return token;
  }

  void commit(String tokenId) {
    final t = store.reservations[tokenId];
    if (t == null) {
      throw NotFoundError('reservation not found: $tokenId');
    }
    if (!t.isOpen) {
      throw ReservationError('reservation $tokenId already closed');
    }
    final s = stockOf(t.sku);
    s.onHand -= t.quantity;
    s.reserved -= t.quantity;
    t.committed = true;
  }

  void release(String tokenId) {
    final t = store.reservations[tokenId];
    if (t == null) {
      throw NotFoundError('reservation not found: $tokenId');
    }
    if (!t.isOpen) {
      throw ReservationError('reservation $tokenId already closed');
    }
    final s = stockOf(t.sku);
    s.reserved -= t.quantity;
    t.released = true;
  }
}
